README.md
[![Build Status](https://img.shields.io/travis/com/imath/entrepot/master.svg)](https://travis-ci.com/imath/entrepot)

# Entrepôt

## Français 🇫🇷
*Inscrivez vos extensions, thèmes et blocs WordPress gratuits dans l'Entrepôt.*

Cette extension WordPress vous propose de recenser vos propres extensions, thèmes et blocs gratuits hébergés sur GitHub.com directement depuis :
- l'onglet **Entrepôt** de l'écran d'administration de l'ajout d'extensions,
- l'onglet **Entrepôt** de l'écran d'administration de l'ajout de thèmes ou depuis la section **Thèmes de l'Entrepôt** de l'interface de changement de thèmes du Customizer de WordPress,
- le menu **Types de bloc** de l'administration de WordPress.

Grâce à l'**Entrepôt** vos dépôts publics hébergés sur GitHub.com bénéficieront, par ailleurs, d'une installation et de mises à jour directement déclenchables depuis l'administration des sites WordPress de vos utilisateurs.

En téléchargeant et en installant cette extension, vos utilisateurs disposent donc d'une source alternative d'extensions, de thèmes et de blocs gratuits.

Les *auteurs* d'extensions, de thèmes ou de blocs, peuvent, à l'aide d'une "Pull Request" particulière référencer les dépôts publics hébergés sur GitHub.com correspondants. Pour en savoir plus, reportez-vous à la [documentation](https://github.com/imath/entrepot/wiki).

Cette extension nécessite WordPress 5.0 et a été testée jusqu'à sa version 5.6.

## English 🇬🇧
*Register your free WordPress plugins, themes and blocks into the Entrepôt*

This plugin lets you list your own free & public GitHub.com hosted plugins themes, and blocks inside:
- the **Entrepôt** tab of the Add Plugins Administration screen,
- the **Entrepôt** tab of the Add Themes Administration screen, or from the **Entrepôt Themes** section of the area of the Customizer to change the active theme,
- the **Block Types** menu of the WordPress Dashboard.

Thanks to the **Entrepôt** your free, public & GitHub hosted repositories will benefit from a one-click installation and updates directly from the WordPress site's Administration of your users.

Users can download this plugin to enjoy an alternative source of free WordPress plugins, themes and blocks.

*Authors* can, using a particular "Pull Request" register their free GitHub.com hosted plugins, themes or blocks. For more information, see the [documentation](https://github.com/imath/entrepot/wiki).

This plugin requires WordPress 5.0 and been tested up to 5.6.

## Code of Conduct

Please take a minute to read the project's [code of conduct](https://github.com/imath/entrepot/blob/master/CODE_OF_CONDUCT.md).


CHANGELOG.md
# Change Log

## 1.5.5

_Requires WordPress 5.0_
_Tested up to WordPress 5.6_

### Bug fixes

- Avoids messing with the Block Editor `query_plugins` requests about block plugins.
- When the WordPress version installed on the site is >= 5.5, disable the feature to manage plugin versions in favor of the WordPress 5.5 manual plugin down/upgrade one.
- Set default values for WP/PHP version theme requirements to make sure it's still possible to install a Theme registered into the Entrepôt using the Customizer.
- Adds PHP/WP requirements management for Plugin updates.
- Adds PHP/WP requirements management for Block updates.

### Newly Registered Plugin

- [Réception](https://github.com/imath/reception/)
- [Communauté protégée](https://github.com/imath/communaute-protegee/)
- [Type de membres](https://github.com/imath/types-de-membre/)

---

## 1.5.4

_Requires WordPress 5.0_
_Tested up to WordPress 5.4_

### Bug fixes

- Fix the appearance of the action buttons used into the Plugin/Theme's modal.
- Fix the "Activités de publication" dependency descriptions into the repository's JSON file.

### Newly Registered Plugin

- [DocuThèques](https://github.com/imath/docutheques/)

---

## 1.5.3

_Requires WordPress 4.8_
_Tested up to WordPress 5.1_

### Bug fixes

- Fix 2 warnings into `Entrepot_REST_Blocks_Controller->prepare_item_for_response()`

### Newly Registered Block

- [Section](https://github.com/imath/section/)

---

## 1.5.2

_Requires WordPress 4.8_
_Tested up to WordPress 5.1_

### Bug fixes

- Check if the hook is a regular function or a class when dealing with admin notices.

### Props

@gregoirenoyelle

---

## 1.5.1

_Requires WordPress 4.8_
_Tested up to WordPress 5.1_

### Bug fixes

- Make sure `entrepot_blocks_dir()` is available even if WordPress version < 5.0
- Do not include `.babelrc` in built package.

### Props

@gregoirenoyelle

---

## 1.5.0

_Requires WordPress 4.8_
_Tested up to WordPress 5.1_

### Features

- Manage standalone Block Types from a specific Administration screen.
  - handle Block Type installations
  - handle Block Type updates
  - handle Block Type activations
  - handle Block Type deactivations
  - handle Block Type deletions
- Introduce a new Entrepôt Block Types API in order to use activated Block Types within the Block editor.
- Introduce a distant way of getting registered repositories using the GitHub API.

### Bug fixes

- Fix PHPUnit7 failings in Travis

### Newly Registered Block

- [Formulaire de Recherche](https://github.com/imath/formulaire-de-recherche/)
- [Entrepôt Test Block](https://github.com/imath/entrepot-test-block/) (Only loaded when `WP_DEBUG` is `true`)

---

## 1.4.2

_Requires WordPress 4.8_

### Bug fixes

- Remove the Gutenberg dependency for GutenBlocks and replace it with WordPress 5.0.

## 1.4.1

_Requires WordPress 4.8_

### Newly Registered Plugins

- Activités de Publication

---

## 1.4.0

_Requires WordPress 4.8_

### Features

- Theme installs/updates inside the customizer and from the regular Administration screens.

### Bug fixes

- Leave WordPress display regular w.org plugin details in the Thickbox.
- Make sure the zip file type check in Plugin overwrites is taking in account all zip mime types (eg: "application/x-zip-compressed").
- Use actions instead of filters when overring update transients.

### Newly Registered Themes

- Vingt DixSept

---

## 1.3.0

_Requires WordPress 4.8_

### Features

- Improve the Admin notifices center by adding a link to fully open notices.
- Update the URL to flag plugins.

### Bug fixes

- Fix a notice error when notice hooks are hooked from class methods.

---

## 1.2.1

_Requires WordPress 4.8_

### Newly Registered Plugins

- Gutenblocks
- WP Tuning

---

## 1.2.0

_Requires WordPress 4.8_

### Bug fixes

- Removes extra parenthesis when using require or require_once.

### Features

- More links to inform about Entrepôt repositories into the "More details" modal
- Use the Entrepôt repository icon into the WordPress Bulk update List Table.
- Manual WordPress plugin upgrade and downgrade.
- Restrictions to the WordPress Plugins Code Editor.

### Newly Registered Plugins

- Alternative Control for Public Group
- Profil de Groupes

---

## 1.1.0

_Requires WordPress 4.8_

### Bug fixes

- Fix a WordPress JavaScript error about Updates count into the Entrepôt Tab of the Install Plugins Administration screen.
- Improve version checking when a plugin is not using a standard version number.
- Multisite improvements to make available the plugin depedencies available for each blog of the network.

### Features

- Plugin dependencies management
- Admin Notices Center
- Upgrades API

### Newly Registered Plugins

- AD ACF Builder
- BP Idea Stream
- BP Reshare

---

## 1.0.0

_Requires WordPress 4.8_

### Features

Thanks to Entrepôt you can enjoy an alternative source of public & free Plugins that are hosted on GitHub.com. Once activated, you will be able to **browse, install, activate and upgrade** the registered plugins directly from your WordPress Administration.

### Registered Plugins

- Entrepôt
- MédiaThèque
- WP Idea Stream
- WP Statuses


entrepot.php
<?php
/**
 * Plugin Name: Entrepôt
 * Plugin URI: https://github.com/imath/entrepot/
 * Description: Une liste d'extensions, de blocs et de thèmes gratuits hébergés sur GitHub.com.
 * Version: 1.5.5
 * Requires PHP: 5.6
 * Requires at least: 5.0
 * Tested up to: 5.6
 * License: GNU/GPL 2
 * Author: imath
 * Author URI: https://imathi.eu/
 * Text Domain: entrepot
 * Domain Path: /languages/
 * Network: True
 * GitHub Plugin URI: https://github.com/imath/entrepot/
 */

// Exit if accessed directly.
defined( 'ABSPATH' ) || exit;

if ( ! class_exists( 'Entrepot' ) ) :
/**
 * Main plugin's class
 *
 * @package Entrepôt
 *
 * @since 1.0.0
 */
final class Entrepot {

	/**
	 * Plugin's main instance
	 *
	 * @var object
	 */
	protected static $instance;

	/**
	 * Initialize the plugin
	 *
	 * @since 1.0.0
	 */
	private function __construct() {
		$this->globals();
		$this->inc();
	}

	/**
	 * Return an instance of this class.
	 *
	 * @since 1.0.0
	 *
	 * @return object A single instance of this class.
	 */
	public static function start() {

		// If the single instance hasn't been set, set it now.
		if ( null == self::$instance ) {
			self::$instance = new self;
		}

		return self::$instance;
	}

	/**
	 * Setups plugin's globals
	 *
	 * @since 1.0.0
	 */
	private function globals() {
		// Version
		$this->version = '1.5.5';

		// Domain
		$this->domain = 'entrepot';

		// Base name
		$this->file     = __FILE__;
		$this->basename = plugin_basename( $this->file );

		// Path and URL
		$this->dir              = plugin_dir_path( $this->file );
		$this->url              = plugin_dir_url ( $this->file );
		$this->js_url           = trailingslashit( $this->url . 'js' );
		$this->assets_url       = trailingslashit( $this->url . 'assets' );
		$this->assets_dir       = trailingslashit( $this->dir . 'assets' );
		$this->inc_dir          = trailingslashit( $this->dir . 'inc' );
		$this->repositories_dir = trailingslashit( $this->dir . 'repositories' );

		// Plugins missing dependencies.
		$this->miss_deps = array();

		// Plugins upgrade tasks.
		$this->upgrades = array();

		// Block supports
		$this->block_supports = function_exists( 'render_block' );
	}

	/**
	 * Includes plugin's needed files
	 *
	 * @since 1.0.0
	 */
	private function inc() {
		spl_autoload_register( array( $this, 'autoload' ) );

		require $this->inc_dir . 'functions.php';
		require $this->inc_dir . 'customizer.php';

		if ( $this->block_supports ) {
			require $this->inc_dir . 'blocks.php';
		}

		if ( is_admin() ) {
			require $this->inc_dir . 'admin.php';
		}

		// Load deprecated functions.
		if ( defined( 'WP_DEBUG' ) && WP_DEBUG ) {
			require $this->inc_dir . 'deprecated.php';
		}

		require $this->inc_dir . 'hooks.php';
	}

	/**
	 * Class Autoload function
	 *
	 * @since  1.0.0
	 *
	 * @param  string $class The class name.
	 */
	public function autoload( $class ) {
		$name = str_replace( '_', '-', strtolower( $class ) );

		if ( false === strpos( $name, $this->domain ) && 'Parsedown' !== $class ) {
			return;
		}

		$path = $this->inc_dir . "classes/class-{$name}.php";

		// Sanity check.
		if ( ! file_exists( $path ) ) {
			return;
		}

		require $path;
	}
}

endif;

/**
 * Boot the plugin.
 *
 * @since 1.0.0
 */
function entrepot() {
	return Entrepot::start();
}
add_action( 'plugins_loaded', 'entrepot', 1 );
